### PR TITLE
release 1.10/ack improvement highcpu consumption

### PR DESCRIPTION
**Pull request by astelmashenko**
Fixes #

Highcpu consumption. Introduced by PR https://github.com/knative-extensions/eventing-natss/pull/267/commits/41b2d7661ba85cb6b8e77ea404179cb7f0a0db37 to solve jetstream redelivery. It should be controlled by AckWait consumer configuration.

cc @dan-j 

## Proposed Changes

Removed code which created timer for each incomming message to mark InProgress to prevent redelivery from JetStream steam.
**Big changes**: implemented retries based on JetStream deliveryMax feature instead of in-memory golang retry module.

- 🐛 Fix bug

**Release Note**

```release-note
Fixes high cpu consumption of dispatcher
```

### CODE REVIEW

**Comment by dan-j** [comment]
So I added this because we ran into an issue where the time to spin up a Knative Service from 0-replicas was longer than the default AckWait. The dispatcher doesn't know when the `*nats.Msg` has expired, and the request is held on the knative activator until the service is ready. By the time the request is actually handled the AckWait has passed and the Ack fails.. However, now JetStream thinks the message hasn't been delivered so attempts a redelivery.

In an ideal scenario, the dispatcher would know about the AckWait and dispatch the actual HTTP request using a context with timeout `AckWait - jitter` (some jitter to make sure there's time for the dispatcher to do it's processing before/after the request). It doesn't look like you can access the AckWait/deadline on the `*nats.Msg` directly, but you could get the `ConsumerInfo` from the `*nats.Subscription` when creating the `Consumer` during `Dispatcher#subscribe()`. The `ConsumerInfo` has a `Config.AckWait` field, but I'd double check this is properly set if defaults are used when it's created.

I wouldn't want to merge this PR as-is until we have a solution for this 0-replica cold-start issue, but happy to hear any alternative solutions?

**Comment by dan-j** [comment]
Ah, another point...

I'm not 100% sure, but I feel like `nats.QueueSubscribe()` receives messages in batches and then calls the `MsgHandler` in serial. I observed this when working on https://github.com/knative-extensions/eventing-natss/commit/41b2d7661ba85cb6b8e77ea404179cb7f0a0db37, which is why forwarding the actual event is done in another goroutine.

If this PR changes to my suggestion of setting a timeout on the request context, the jitter also needs to take into account that, from an `AckWait` perspective, the message is considered as "delivered" the moment it's received by the client. I'm not sure there's a way to know what this duration is, I imagine it will typically be sub-milliseconds but could spike to longer if there's high load.

**Comment by astelmashenko** [comment]
> So I added this because we ran into an issue where the time to spin up a Knative Service from 0-replicas was longer than the default AckWait. The dispatcher doesn't know when the *nats.Msg has expired, and the request is held on the knative activator until the service is ready. By the time the request is actually handled the AckWait has passed and the Ack fails.. However, now JetStream thinks the message hasn't been delivered so attempts a redelivery.
Regarding this part, you can set AckWait bigger and set request timeout. E.g. it take 20sec to spin up a service and 30 seconds to process a request, then you set delivery.timeout to 35sec and Channel AckWait to 35*(number of retries)+delta, e.g. 120sec

**Comment by astelmashenko** [comment]
@dan-j ,  overall as I understand you need to calculate AckWait based on delivery.retry and delivery.timeout you set to you channel's subscription. What is the problem just to set it big enough, e.g. 300s? and control slow consumers by MaxAckPending.

Having that timer you basiclly set AckWait to unlimited and add additional load on cpu. In my case I have broker and around 30 triggers which leads to 30 consumers with MaxAckPending=1000 it becomes 30000 timers in the worst case, it creates very big load on CPU.

**Comment by astelmashenko** [comment]
> I'm not 100% sure, but I feel like nats.QueueSubscribe() receives messages in batches and then calls the MsgHandler in serial.

I'm not sure about that, it is push based consumer, if I scale dispatcher it round-robin messages one by one in turn.

**Comment by dan-j** [comment]
> Regarding this part, you can set AckWait bigger and set request timeout. E.g. it take 20sec to spin up a service and 30 seconds to process a request, then you set delivery.timeout to 35sec and Channel AckWait to 35*(number of retries)+delta, e.g. 120sec

Yeah this is what I'm getting at. You could do that, but this PR would still introduce bugs in our environment, because the request dispatcher doesn't set any timeout/deadline on the request's `context.Context`. Which is what I mean about not wanting to merge this as-is.

I'm happy to remove the ticker, your issue is a real one which I agree needs fixing, just we need a solution which works for both of us.

One concern I've always had with the operator is how we handle retries. At the moment it's really confusing because we retry in multiple places: 1) at the JS layer via a consumer's `maxDelivery` and 2) at the in-memory layer via the `DispatchMessageWithRetries()` function which configured by the `Subscription`.

It would be nice to solve both issues here.

The simplest option is to remove redelivery from the JS layer and reuse the `DispatchMessageWithRetries()` functionality. However this would result in message loss in the event of a pod failure.

The more robust solution is to remove `NatsJetStreamChannel.spec.consumerConfigTemplate.ackWait` and `NatsJetStreamChannel.spec.consumerConfigTemplate.maxDeliver` from the CRDs, and calculate the proper JS consumer options based on the `Subscription.spec.delivery` configuration. Then let JS do all the retrying and use `DispatchMessage` instead of `DispatchMessageWithRetries`, setting the `context.Context` properly to abort requests which time out. This might be a bit of a headache to do exponential backoffs because ackWait would need to be the maximum possible, and the context's timeout would need to be calculated on each retry (which is possible because you can get the redelivery counter from `meta, _ := msg.Metadata(); meta.NumDelivered`)

**Comment by astelmashenko** [comment]
hey @dan-j , I've done initial implementation (have not tested it though), your review is welcomed.

File: pkg/channel/jetstream/dispatcher/consumer.go
Comment (dan-j) [inline pkg/channel/jetstream/dispatcher/consumer.go:81]: This is looking really good now, my only concern is how we cancel requests when the AckWait expires, as this was the main problem for us which was solved by the ticker goroutine.

The messages are buffered in the client before calling our `MsgHandler` method, so simply creating a `context.WithTimeout(ctx, config.AckWait)` may actually result in a message being handled more than once. I was hoping the `msg.Metadata().Timestamp` would be the time from which the ack-wait starts from, but this is the time the message was added to the stream, and doesn't change with redeliveries.

I think you might have to do something like:

```go
meta, _ := msg.Metadata()
// probably store this as a struct field on the Consumer
info, _ := c.jsSub.ConsumerInfo()

// if each delivery has timed out, then multiplying the number of deliveries by the ack wait will give us the
// duration from publish which this attempt will be ack-waited
ackDurationFromPublish := time.Duration(meta.NumDelivered) * info.Config.AckWait
// add a bit of jitter
jitter := 500 * time.Millisecond

// the deadline is the published timestamp plus our duration calculated above
deadline := meta.Timestamp.Add(ackDurationFromPublish).Add(-jitter)

// if previous deliveries were explicitly nacked earlier than the deadline, then our actual deadline will be earlier
// than the deadline above
ackDeadlineFromNow := time.Now().Add(info.Config.AckWait).Add(-jitter)
if deadline.After(ackDeadlineFromNow) {
	deadline = ackDeadlineFromNow
}

ctx, cancel := context.WithDeadline(ctx, deadline)
defer cancel()

result = c.doHandle(ctx, msg)

...
```

**Review by dan-j — CHANGES_REQUESTED** [review]
Just the one comment around how we cancel requests but otherwise LGTM

**Comment by dan-j** [comment]
Just to add to the above. We've load tested both the current version and this PR and definitely recognise the performance improvements so I'm happy to merge once my one comment above is addressed.

However, I believe there's still more to look into regarding performance, we're noticing a steady increase in CPU over time when put under continuous load, but memory use remains stable, which is a bit confusing. After some thought, our theory is it's on the NATS client itself, but requires us to do some profiling to know for sure. 

But, this should be a new issue and not block this PR. I can create a new issue once we have some more concrete evidence, it could just be a configuration thing we need to document, or an issue on NATS itself.

File: pkg/channel/jetstream/dispatcher/consumer.go
Comment (dan-j) [reply pkg/channel/jetstream/dispatcher/consumer.go:81]: The jitter should also probably be quite a bit lower, or a fraction of AckWait, because I imagine some deployments may have a small AckWait where 500ms may actually cause more problems than they solve.

Comment (dan-j) [reply pkg/channel/jetstream/dispatcher/consumer.go:81]: So, I've continued some more reading and now I'm at the conclusion we're building the consumers wrong. I'm happy to take this into a new issue though, if you think that's appropriate.

We're using push consumers, and reading this page, they're described as legacy: https://natsbyexample.com/examples/jetstream/push-consumer/go

I'm not sure who the author of natsbyexample.com is, but it's spot on with the issues which arise when consumers can't keep up.

We should probably refactor where consumers are created and use a pull consumer instead, using `FetchNoWait` to process as many as possible in a batch size we think is reasonable: https://natsbyexample.com/examples/jetstream/pull-consumer/go

What do you reckon @astelmashenko, address here or a new issue?

Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/consumer.go:81]: pull consumer looks more reasonable to me, because it allows to better control back-pressure. I'm not sure about how to controll back-pressure in golang world with goroutines, how to control max inflight requests?

Regarding the comment with deadline calculation, it looks complex and I'd just use 
`ackDeadlineFromNow := time.Now().Add(info.Config.AckWait).Add(-jitter)`
I thinkg probability of duplicate message in case of timeout will be almost the same.

But I'll add all those calculations because it is already implemented.

Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/consumer.go:81]: > This is looking really good now, my only concern is how we cancel requests when the AckWait expires, as this was the main problem for us which was solved by the ticker goroutine.

It is covered already by jitter, see defaults.go
`consumerConfig.AckWait = retryConfig.RequestTimeout + jitter`
and I calculate context timeout as just `retryConfig.RequestTimeout`
`ctx, cancel = context.WithTimeout(ctx, c.sub.RetryConfig.RequestTimeout)`

so AckWait is bigger then RequestTimeout and there is jitter time to Ack message if request almost reached timeout.
Maybe jitter should be lower, I set it 5sec, maybe just change it to 500ms.

Example:
RequestTimeout = 30s 
AckWait = 30s + 500ms

case1: request completed in 2s, it's ok message Ack'ed
case2: request timed out, context cancelled in 30s, there is 500ms to NaK message, even if msg is not NaK'ed it will be redelivered in any case
case3: request completed in 29,999s, there is 500ms to Ack message

the only concern is the last case3, IMO probability of such case is pretty low, at least because requestTimeout should be bigger then p95 of request processing time by a target service.

Should we do all those complex calculations? do you still see it reasonable?

Comment (dan-j) [reply pkg/channel/jetstream/dispatcher/consumer.go:81]: So you're correct, but the reason we added the ticker was to address case 3 because this was happening a lot if the subscriber is scaling up and the request gets queued in knative's activator. We were observing the message being pending on the NATS client for 10-20 seconds before our MsgHandler is called.

I see two options:

1. Move to pull consumer and do away with my obscure maths
2. Do my obscure maths and create a new issue to move to pull consumer, then we'll remove this as part of that issue

**Comment by astelmashenko** [comment]
@dan-j , here are results of load test. Scenario is pretty simple: 
http workers -> broker (channel) -> trigger (sub) -> functionQ
A functionQ processes a request with random sleep.
Trigger has request timeout, so that functionQ 25% of times fails to process in time.
Please pay attention on blue line, it is http latency of the dispatcher.
Retries are in-memory:
image
Retries are based on JetStream:
image

**Comment by dan-j** [comment]
Awesome stuff! Thanks for the graphs too 👍 

It's a shame we've had to reimplement the whole message_dispatcher logic but from our earlier conversations this was going the be the only way.

I'm happy to merge this, but need to fix my permissions on the github org so I can't do it just yet

**Comment by dan-j** [comment]
Codecov checks are failing:
- https://github.com/knative-extensions/eventing-natss/pull/426/checks?check_run_id=18216441538
- https://github.com/knative-extensions/eventing-natss/pull/426/checks?check_run_id=18216441749

**Comment by astelmashenko** [comment]
@dan-j , now it is green , ready to merge

**Comment by dan-j** [comment]
@pierDipi could we have this merged? I will try to sort my permissions on knative/org or wherever it is this week

**Comment by creydr** [comment]
Hi @astelmashenko,
is there a reason, why you're targeting directly 1.10 branch and not main and then backporting it?

**Comment by astelmashenko** [comment]
Because we are using it in production. I'll backport to later versions with e.g. cherry-pick or manually.

**Comment by dan-j** [comment]
@astelmashenko can we update the PR to go into `main` and then cherry-pick afterwards.

**Comment by astelmashenko** [comment]
@dan-j , I would not do that, main was updated and, afaik, is not directly compatible with this PR

**Comment by dan-j** [comment]
Ah, fair enough. Let's try and get this all wrapped up this week. I've created knative/community#1479 to add us both as approvers.

@zhaojizhuang at the moment you're the only approver of this repo, or there's probably an eventing admin who can look.

Once this is in, I'll get the changes onto latest too

**Comment by dan-j** [comment]
@astelmashenko and I are now approvers, but since this is being merged into [knative-extensions:release-1.10](https://github.com/knative-extensions/eventing-natss/tree/release-1.10) the `OWNERS_ALIASES` file isn't updated.

Can someone from @knative-extensions/eventing-writers please approve this PR?

**Comment by pierDipi** [comment]
@dan-j we can backport the approvers update, feel free to open a PR

**Pull request merged**
